src/tree.rs

```diff
@@ -126,6 +128,18 @@ pub fn minimum_spanning_tree(
     let mut spanning_tree = (*graph).clone();
     spanning_tree.graph.clear_edges();
 
+    _minimum_spanning_tree(py, graph, &mut spanning_tree, weight_fn, default_weight)?;
+    Ok(spanning_tree)
+}
+
+/// Helper function to allow reuse of spanning_tree object to reduce memory allocs
+fn _minimum_spanning_tree<'a>(
+    py: Python,
+    graph: &graph::PyGraph,
+    spanning_tree: &'a mut graph::PyGraph,
+    weight_fn: Option<PyObject>,
+    default_weight: f64,
+) -> PyResult<&'a mut graph::PyGraph> {
```

Comment:
Sorry, I misread the code here when I left the comment. Looking at the changes you just made I didn't realize that making this change would require changing the `bipartition_tree` function to return the spanning tree so that `bipartition_graph` could use the graph on each iteration. When looking at `bipartition_tree` as a standalone python function now it's api seems less than ideal because of that tuple return type.

When I left this comment earlier I was only concerned about the `minimum_spanning_tree()` -> `_minimum_spanning_tree()` call path and it looked odd to me to have a function that takes a mutable reference and then returns a mutable reference to the same object. Normally when you require a mutable borrowed input parameter you don't need to return that reference as part of the function signature because the caller already has the same reference.Maybe instead of passing by value here you can restore the mutable reference here and just change the return type to be `PyResult<()>` so that we return errors. That would retain the user facing pyfunction api for `bipartition_tree` but also avoid the double reference. The other thing I was thinking is maybe make an internal `_bipartition_tree` that works with rust native types so we can develop the python interface in isolation from the internal rust calling of that function. But, I'll defer to what pattern you think is best here as you're the first/primary user of these functions so you're in the best position to evaluate that making sure the API we end up with here is usable.

Reply:
Ok, I changed the return type to be `PyResult<()>`, as per your suggestion. I'd like to have `bipartition_tree` exposed, since sometimes we want to use Wilson's Uniform Spanning Tree algo or other spanning tree methods (active area of research), which we already have implemented in Python, so keeping the ability to pass a custom spanning tree would be nice.

If alternative spanning tree methods become more widely used in real-world redistricting efforts, I may want to port them to Rust/retworkx as well. I've renamed `bipartition_tree` to `bipartition_tree_mst` to reflect the possibility that other spanning tree methods may be useful to have in the future.

Reply:
Oh, I just meant that if it made it easier to work with you could have `bipartition_tree()`/`bipartion_tree_mst()`  and `bipartition_graph()` call `_bipartition_tree()` if that gave you more flexibility around the internal rust types (because we wouldn't need to worry about having types that directly have `FromPyObject` implemented). But it's not strictly necessary, just me thinking out loud.